rlog/2024-07-01-bloomfilter.mdx

```diff
@@ -0,0 +1,142 @@
+However, this assumption can be removed to gain the same approximation.
+
+We note that for a single hash function, the probability that a specific bit is flipped to 1 is $1/n$.
+So, the probability that the specific bit is not flipped by the hash function is $1-1/n$.
+Applying our assumption that the $k$ hash functions are 'independent,'
+the probability that the specific bit is not flipped by any of the hash functions is
+$(1-1/n)^k$.
+
+Recall the Calculus fact $\lim_{\infty} (1-1/n)^n = e^{-1}$.
+That is, as we increase the number of bits that our Bloom filter uses, the approximate probability that a given bit is not flipped by any of the $k$ hash functions is $e^{-k/n}$.
+
+Suppose that $\ell$ entries have been added to the Bloom filter.
+The probability that a specific bit is still 0, after the $\ell$ entries have been added, is approximately $e^{-\ell k/n}$.
+The probability that a queried element is erranously claimed as a member of the digest is approximately
+$(1-e^{-\ell k/n})^k$.
+
+The following table provides a concrete values for these approximations.
+
+
+
+| $n$ | $k$ | $\ell$ | $(1-e^{-\ell k/n})^k$|
+| -------- | -------- | -------- | --- |
+| 32     |   3   |  3    | 0.01474 |
+| 32     |   3   |  7    | 0.11143 |
+| 32     |   3   |  12    |0.30802 |
+| 32     |   3   |  17    |0.50595 |
+| 32     |   3   |  28    |0.79804 |
+
+Notice that the probability of false-positives increases as the number of elements ($\ell$) have been added to the digest increases.
+
+### Sliding-Window Bloom Filter
+Our toy example and table illustrated an issue concerning Bloom filters.
+The number of entries that can be added to a Bloom filter is restricted by our choice of $k$ and $n$.
+Not only is does the probability that false-positives will occur increase,
+but it is possible that our vector ${\bf{v}}$ can a string of all 1's.
+[Szepieniec and Værge](https://eprint.iacr.org/2023/1208.pdf) proposed a modification to Bloom filters to handle this.
+
+Instead of having a fixed number of bits for our Bloom filter, we dynamically allot memory based on the number of entries that have been added to the filter.
+Given a predetermined threshold ($b$) for the number of entries, we shift our 'window' of flipping bits by $s$ bits.
+Note that this means that it is necessary to keep track of when a given entry is added to the digest.
+This means that querying the sliding-window Bloom filter will yield different results when different timestamps are used.
+
+This can be done with $k$ hash functions as we used earlier.
+Alternatively, Szepieniec and Værge proposed using the same hash function but to produce $k$ entries in the current window.
+Specifically, we obtain the bits we wish to flip to 1's for entry $x$ by computing $h(x || i)$ for each $i \in \{0,\dots, k-1\}$.
+
+By incorporating a shifting window, we maintain the efficient query and appending at the cost of constant space.
+However, by losing constant space, we gain 'infinity' scalability.
```

Comment:
err, maybe this is a catch 22, the infinite scalability will make it more difficult for someone to quickly verify the existence of members in the set

Reply:
I've added the algorithms for append/query here to make this clear.

Reply:
awesome